Opswat filescan

anikobartos opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
There is no issue. This is a new integration pack.

## Description
This is an integration pack for OPSWAT Filescan. It implements the following functions:
- scan url
- scan file
- search with query-string

## Screenshots
![image](https://user-images.githubusercontent.com/75678714/228381103-10659bd7-fd31-488f-939b-04bacd25df3d.png)
![image](https://user-images.githubusercontent.com/75678714/228381178-ec395b1e-be45-4aa0-b3e1-0617f5ab3549.png)
![image](https://user-images.githubusercontent.com/75678714/228381230-a3dde315-9c97-46ae-b08f-0fa036d015ae.png)
![image](https://user-images.githubusercontent.com/75678714/228381344-30db70e4-84e2-44a8-8787-44610bd20fa0.png)
![image](https://user-images.githubusercontent.com/75678714/228381490-3737e0d4-2ff9-485a-a64a-4d21479aeb99.png)


## Minimum version of Cortex XSOAR
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=25581) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=25581) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=25581) it.</sub>

anikobartos commented:
Hi,
I have two questions:
- Do I need to add a release notes file? I have an error because it's missing, but the documentation says that it's not needed on the first release.
- The second is that I've got an error at the "Run Unit Testing And Lint" check (see below), however I can't figure out what the problem could be, as the error message is quite brief. Could you please help me out?
```
#######################
 Image creation errors 
#######################
  - Package: OPSWATFilescan
    Image: demisto/py3-native:8.2.0.52641
      Error: Installation script failed to run on container
             '5f8175340a757f9d095c062f9a1fb0549b2315690e8e7801a4b1606484e1b7f7'.
```

dantavori commented on Packs/OPSWAT-Filescan/pack_metadata.json:
```diff
@@ -0,0 +1,20 @@
+{
+    "name": "OPSWAT Filescan",
+    "description": "Unique adaptive threat analysis technology",
+    "support": "xsoar",
```
```suggestion
    "support": "partner",
```
I suppose this is a mistake and this contribution is supported by you - if this is not the case, please let us know.

anikobartos replied:
Thanks! You are right. I corrected the pack_metadata file

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,390 @@
+        return self._http_request(
+            method="GET",
+            ok_codes=([200]),
+            params={"query": query_str, "page_size": 10, "page": 1},
```
1) what's the reason we always requesting for page 1? 
2) what's the reason we always requesting for page_size 10?

anikobartos replied:
In this first version of the integration, with the search request, we always give back the first 10 responses from the 1st page.    (This may change in future versions.)

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,390 @@
+        score = Common.DBotScore.BAD
+    elif verdict.upper() == "SUSPICIOUS":
+        score = Common.DBotScore.SUSPICIOUS
+    else:
+        score = Common.DBotScore.NONE
```
```suggestion

```

not needed if you set `score = Common.DBotScore.NONE` at line 110.

anikobartos replied:
Else part removed. Thanks the comment!

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,390 @@
+            score = Common.DBotScore.BAD
+        elif verdict.upper() == "SUSPICIOUS":
+            score = Common.DBotScore.SUSPICIOUS
+        else:
+            score = Common.DBotScore.NONE
```
```suggestion
```

anikobartos replied:
else part removed as well

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,390 @@
+def build_reputation_result(api_reponse: Dict[str, Any]):
+    reports = api_reponse.get("reports", {})
+
+    if len(reports) == 1:
+        first_report = next(iter((api_reponse.get("reports", {}).items())))[1]
+        return build_one_reputation_result(first_report)
```
```suggestion

```

i think you can remove it and handle the case where there is one report in lines 275-279 as well.
see suggested implementation below

anikobartos replied:
answer is under the other comment below

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,390 @@
+    )
+
+
+def build_reputation_result(api_reponse: Dict[str, Any]):
+    reports = api_reponse.get("reports", {})
+
+    if len(reports) == 1:
+        first_report = next(iter((api_reponse.get("reports", {}).items())))[1]
+        return build_one_reputation_result(first_report)
+
+    elif len(reports) > 1:
+        command_res_ls = []
+        for report in reports:
+            command_res_ls.append(build_one_reputation_result(reports[report]))
+        return command_res_ls
```
```suggestion
def build_reputation_result(api_reponse: Dict[str, Any]):
    reports = api_reponse.get("reports", {})
    if isinstance(reports, dict):  # there is only a single report
        reports = [reports]
    
    command_res_ls = []
    for report in reports:
        command_res_ls.append(build_one_reputation_result(reports[report]))
     return command_res_ls
```

maybe something like that?

anikobartos replied:
Actually, `reports` is already a list, so it doesn't necessary to put it to another list like this: reports = [reports]
I simplified the code and corrected the relevant test case as well.

GuyAfik replied:
in that case make sure the default value will be `list` if it does not exist:
`reports = api_reponse.get("reports", [])`

anikobartos replied:
Corrected!

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,390 @@
+    results = client.test_module()
+    if "accountId" in results:
+        return "ok"
+    raise DemistoException("\nTest module failed, {}".format(results))
```
```suggestion
    raise DemistoException(f"\nTest module failed, {results}")
```

use f-strings :)

anikobartos replied:
Thanks! Corrected!

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,390 @@
+        else:
+            raise NotImplementedError(f"{command} command is not implemented.")
+    except Exception as e:
+        err_msg = f"Exception in {INTEGRATION_NAME} : Failed to execute {command} command: [{e!r}]"
```
```suggestion
        err_msg = f"Exception in {INTEGRATION_NAME} : Failed to execute {command} command: [{e}]"
```

what is `!r`?

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan_test.py:
```diff
@@ -0,0 +1,248 @@
+
+    polling_args = {"flow_id": "1234", "hide_polling_output": True, "continue_to_poll": True, "url": "test.com"}
+    mocker.patch.object(client, '_http_request', return_value=raw_response)
+    # mocker.patch.object(client, "get_scan_result", return_value=raw_response)
```
```suggestion
```

anikobartos replied:
Removed the comment.

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.yml:
```suggestion
          description: The file name.
```

anikobartos replied:
Corrected the description.

GuyAfik reviewed (commented):
wonderful work! i gave some small comments, once those comments are addressed, we can conduct a demo.
you can remove the `Pipfile` and `Pipfile.lock`, those are not needed.

GuyAfik commented:
@anikobartos looks good to me, i would like to do a demo on that if possible, i would appreciate if you can contact me on slack on Demisto DFIR Community (https://www.demisto.com/community/) to setup a demo.

GuyAfik commented:
demo summary:

1) add prefix of opswat to each command
2) make sure all the context of all commands will be under the `OPSWAT` context key.
3) add pagination to `filescan-scan-file` command and remove the print only the summary.

anikobartos commented:
Changes requested based on the demo have been made:
- added pagination to the search query
- renamed commands to opswat-filescan-*
- removed DBotScore part from search-query outputs

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,400 @@
+
+def search_query_command(client: Client, args: Dict[str, Any]):
+    def validate_args():
+        if page_size and (not page_size.isdigit() or int(page_size) not in [5, 10, 20]):
```
what's the reason we pre-define `page_size` to these values only?

GuyAfik replied:
got it, thanks.

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,400 @@
+
+    items = []
+    query_string = args.get("query", "")
+    page_size = args.get("page_size")
+    page = args.get("page")
+    limit = args.get("limit", 10)
```
```suggestion
    page_size = arg_to_number(args.get("page_size"))
    page = arg_to_number(args.get("page"))
    limit = arg_to_number(args.get("limit", 10))
```

use `arg_to_number` from CommonServerPython

anikobartos replied:
Thanks! Fixed.

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,400 @@
+    else:
+        page_size = 20
+        page = 1
+        limit = int(limit)
```
```suggestion
```

will already be `int` from the `arg_to_number`

anikobartos replied:
I removed the int() parts where arg_to_number handle it

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,400 @@
+        continue_query = True
+        while continue_query:
+            actual_items = client.get_search_query(query_string, page, page_size).get("items", [])
+            if len(actual_items) < 20 or len(items) >= limit:
```
what's the reason for `len(actual_items) < 20`?

GuyAfik commented on Packs/OPSWAT-Filescan/Integrations/OPSWATFilescan/OPSWATFilescan.py:
```diff
@@ -0,0 +1,400 @@
+            actual_items = client.get_search_query(query_string, page, page_size).get("items", [])
+            if len(actual_items) < 20 or len(items) >= limit:
+                continue_query = False
+            items += actual_items
```
```suggestion
            items += actual_items
            page += 1
```

don't you need to raise the page each iteration?

GuyAfik reviewed (commented):
@anikobartos wonderful work, see my comments, for any questions feel free to slack me.

GuyAfik reviewed (approved):
@anikobartos too easy, wonderful work!
